### ISSUE TITLE
WASM's FPDFBitmap_GetBuffer implementation is unsound, causing overwritten image data

### ISSUE THREAD

**Issue by samsieber** [ajrcarey/pdfium-render#174]
Details:
* pdfium-render = { version = "0.8.26", default-features = false, features = ["thread_safe", "image_025", "image", "pdfium_6666"]}
* Using the WASM build from https://github.com/paulocoutinhox/pdfium-lib/releases/tag/6684
* Reproduction repo: https://github.com/samsieber/pdfium-render-text-garbage/tree/bad-pixels

This is something I've bumped up against occasionally, and wasn't able to come up with a good reproduction test case until now.

Sometimes, when rendering a PDF, I get garbage data for the image, seemingly random white noise at times. I've seen that often rerendering an image will clear that away. I'm not sure what's going on.

I've attached a reproduction test case - it's the same repo as an earlier ticket, but a different branch. 

The test case defaults to pulling a bitmap for a color image out of a PDF. I have a single method that takes the image url "color.pdf" and 6 canvas objects, then using the canvas API it loads the document, parses it, pulls out the bitmap as a DynamicImage and then renders the image to a canvas. So it does that full process 6 times, one for each canvas. 

With the color document, the render starts off with practically nothing, then starts various amounts of the image in subsequent renders. I've also included a black-and-white image. That one renders pure white noise in the first render, and then mostly correct for the rest of the images (there's an issue regarding stride that I'm planning on submitting a patch for grayscale images, which is why that image comes out slanted.

I've attached what those two scenarios look like. 

Something to note is that in my test, the behavior goes away if I make slight tweaks. The smallest one that I think would be useful for debugging is the following function that I wrote to pull out images. I've removed parts that don't change the end result, but note that I was checking bounds to see I could pull the bitmap right out of the PDF without needing to render the full pdf.

But in this function, if I remove the `let bounds = image.bounds()?;` line, the image starts rendering correctly. But also, if I move the rendering to the canvas object into a function, it works correctly. So I'm thinking that perhaps there's an optimization issue in play, or perhaps just memory management. 

```rust
pub fn extract_page_image(page: &PdfPage) -> Result<Option<DynamicImage>, PdfiumError> {
    if page.text().unwrap().is_empty() && page.annotations().is_empty() {
        let images = page.objects().create_group(|_o|  true)?;
        if images.len() == 1 {
            let obj = images.iter().next();
            if let Some(image) =  &obj {
                if let Some(image) = image.as_image_object() {
                    let bounds = image.bounds()?;
                    let media_box = page.boundaries().media()?;
                    return Ok(Some(image.get_raw_image()?))
                }
            } 
        }
    }
    Ok(None)
}
```

Here's an a snippet of the top level function that renders 6 times to different canvases

```rust
#[cfg(target_arch = "wasm32")]
#[wasm_bindgen]
pub async fn write_image_data_for_page(
    url: String,
    index: PdfPageIndex,
    canvas1: HtmlCanvasElement,
    canvas2: HtmlCanvasElement,
    canvas3: HtmlCanvasElement,
    canvas4: HtmlCanvasElement,
    canvas5: HtmlCanvasElement,
    canvas6: HtmlCanvasElement,
) {
    let pdfium = Pdfium::default();

    // ------- Render to the first canvas -----------
    let document1 = pdfium.load_pdf_from_fetch(&url, None).await.unwrap();
    let page1 = document1.pages()
        .get(index)
        .unwrap();
    let dynamic_image = extract_page_image(&page1).unwrap().unwrap();

    let width = dynamic_image.width();
    let height = dynamic_image.height();
    let raw_pixels = dynamic_image.into_rgba8().into_raw();
    let new_img_data = ImageData::new_with_u8_clamped_array_and_sh(
        Clamped(&raw_pixels),
        width,
        height,
    ).unwrap();

    let ctx = canvas1
        .get_context("2d")
            .unwrap()
            .expect("Could not get 2d rendering context for OffscreenCanvas")
            .dyn_into::<CanvasRenderingContext2d>()
            .unwrap();

    canvas1.set_width(width);
    canvas1.set_height(height);

    ctx.put_image_data(&new_img_data, 0.0, 0.0);

    // ------- Render to the second canvas -----------
    let document2 = pdfium.load_pdf_from_fetch(&url, None).await.unwrap();
    // ....
```

![Screenshot 2024-11-18 at 1 21 55 PM](https://github.com/user-attachments/assets/867aaa50-a490-4a8a-a886-d58cf6fd17ea)
![Screenshot 2024-11-18 at 1 22 38 PM](https://github.com/user-attachments/assets/8b73a16a-1c12-48d8-947e-38882aee4b69)

**Comment by samsieber** [comment]
I decided to pursue a narrower test case of just the first render from the second set - the one with the nice repeating static. On a hunch I decided to take each pixel and treat it as a single gray byte, instead of an RGBA, and then tried to parse it as ascii. I get this data repeating:

"FPDFBitmap_GetFormat" [235] "h " NUL

So that's the bytes of the first string in ascii (minus the quotes), then the byte 235, then the bytes for an 'h' and a space in ascii, and then a null byte. Repeating through the whole of the image. 

I've set that up in the pad-pixels two branch of my reproduction repository: https://github.com/samsieber/pdfium-render-text-garbage/tree/bad-pixels . Note that commenting out `src/extract.rs:13` (which only is a call to read the image bounds!) removes this behavior. The reproduction logs out the first 2000 bytes parsed to the console at the info level. I also decided to commit the generated wasm and js file, in case it compiles differently for you. So If you check out this branch you should be able to just use `./serve.sh` and start poking at the example.

![Screenshot 2024-11-20 at 8 29 00 PM](https://github.com/user-attachments/assets/5e2cecf4-ef78-4c15-b0dc-9c01400e113a)

**Comment by samsieber** [comment]
I've done a bit more digging, and it appears that the malloc is behaving oddly. It appears that the buffer in this call overlaps with the vec that is allocated during the to_vec method:

```
 PdfBitmapFormat::Gray => {
                GrayImage::from_raw(width as u32, height as u32, buffer.to_vec())
                    .map(DynamicImage::ImageLuma8)
            }
```            

This discovery came after compiling the reproduction in debug mode (adding `--debug` to the wasm-pack invocation) and adding a panic hook. I got this when running the example.

```
pdfium_render_text_garbage.js:469 panicked at library/core/src/panicking.rs:220:5:
unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap

Stack:

Error
    at http://localhost:4000/pdfium_render_text_garbage.js:472:17
    at logError (http://localhost:4000/pdfium_render_text_garbage.js:238:18)
    at imports.wbg.__wbg_new_abda76e883ba8a5f (http://localhost:4000/pdfium_render_text_garbage.js:471:62)
    at pdfium_render_text_garbage.wasm.console_error_panic_hook::Error::new::h0bf445ad67e4c6ea (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[3174]:0x19a2a6)
    at pdfium_render_text_garbage.wasm.console_error_panic_hook::hook_impl::h1444539405f3d713 (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[1049]:0x13a770)
    at pdfium_render_text_garbage.wasm.console_error_panic_hook::hook::h12db0d38c65beb8e (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[4325]:0x1adac5)
    at pdfium_render_text_garbage.wasm.core::ops::function::Fn::call::h49ede4741d818d1e (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[3859]:0x1a7137)
    at pdfium_render_text_garbage.wasm.std::panicking::rust_panic_with_hook::h6731baa78621a747 (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[1673]:0x16893f)
    at pdfium_render_text_garbage.wasm.std::panicking::begin_panic_handler::{{closure}}::hb6cd8464ed39ae71 (http://localhost:4000/pdfium_render_text_garbage_bg.wasm:wasm-function[2068]:0x17b59b)
    at pdfium_render_text_garbage.wasm.std::sys_common::backtrace::__rust_end_short_backtrace::hbdf3ddeb21a1e747 

(anonymous)	@	pdfium_render_text_garbage.js:469
logError	@	pdfium_render_text_garbage.js:238
imports.wbg.__wbg_error_f851667af71bcfc6	@	pdfium_render_text_garbage.js:466
$console_error_panic_hook::error::he03bace55b652962	@	pdfium_render_text_g…ge_bg.wasm:0x18202e
$console_error_panic_hook::hook_impl::h1444539405f3d713	@	pdfium_render_text_g…ge_bg.wasm:0x13a85c
$console_error_panic_hook::hook::h12db0d38c65beb8e	@	pdfium_render_text_g…ge_bg.wasm:0x1adac5
$core::ops::function::Fn::call::h49ede4741d818d1e	@	pdfium_render_text_g…ge_bg.wasm:0x1a7137
$std::panicking::rust_panic_with_hook::h6731baa78621a747	@	pdfium_render_text_g…ge_bg.wasm:0x16893f
$std::panicking::begin_panic_handler::{{closure}}::hb6cd8464ed39ae71	@	pdfium_render_text_g…ge_bg.wasm:0x17b59b
$std::sys_common::backtrace::__rust_end_short_backtrace::hbdf3ddeb21a1e747	@	pdfium_render_text_g…ge_bg.wasm:0x1b2a0b
$rust_begin_unwind	@	pdfium_render_text_g…ge_bg.wasm:0x1a9844
$core::panicking::panic_nounwind_fmt::h6dbc36adcb815595	@	pdfium_render_text_g…ge_bg.wasm:0x190cb4
$core::panicking::panic_nounwind::h16e38f5a9ff4ae2c	@	pdfium_render_text_g…ge_bg.wasm:0x1a81d5
$core::intrinsics::copy_nonoverlapping::precondition_check::h01c1b2821054a539	@	pdfium_render_text_g…ge_bg.wasm:0x172870
$<T as alloc::slice::hack::ConvertVec>::to_vec::h5ecd5cf0204706d2	@	pdfium_render_text_g…ge_bg.wasm:0x1363d0
$alloc::slice::<impl [T]>::to_vec::h785a73804c9c8afc	@	pdfium_render_text_g…ge_bg.wasm:0x1a9183
$pdfium_render::pdf::document::page::object::image::PdfPageImageObject::get_image_from_bitmap::hf0888be173f1553b	@	pdfium_render_text_g…age_bg.wasm:0xd2f73
$pdfium_render::pdf::document::page::object::image::PdfPageImageObject::get_raw_image::hbce8a67792023678	@	pdfium_render_text_g…ge_bg.wasm:0x13621c
$pdfium_render_text_garbage::extract::extract_page_image::hb4f0536d5085b9b6	@	pdfium_render_text_g…bage_bg.wasm:0xf888
$pdfium_render_text_garbage::write_image_data_for_page::{{closure}}::ha512aac109414d75	@	pdfium_render_text_g…age_bg.wasm:0x1e95f
$pdfium_render_text_garbage::_::__wasm_bindgen_generated_write_image_data_for_page::{{closure}}::h97f6c704d97e04b8	@	pdfium_render_text_g…ge_bg.wasm:0x1154fa
$wasm_bindgen_futures::future_to_promise::{{closure}}::{{closure}}::ha1f67fd213135c0e	@	pdfium_render_text_g…age_bg.wasm:0xf5263
$wasm_bindgen_futures::task::singlethread::Task::run::h1dcc62eb0ac06dff	@	pdfium_render_text_g…ge_bg.wasm:0x122568
$wasm_bindgen_futures::queue::QueueState::run_all::h08835f32d4e10fdf	@	pdfium_render_text_g…ge_bg.wasm:0x1170f2
$wasm_bindgen_futures::queue::Queue::new::{{closure}}::h56253c8cf8ead634	@	pdfium_render_text_g…ge_bg.wasm:0x186742
$<dyn core::ops::function::FnMut<(A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::h7a562a5462e34a46	@	pdfium_render_text_g…ge_bg.wasm:0x1685b4
__wbg_adapter_30
```

If I change that first code to this, manually allocating the new vec:
```
            PdfBitmapFormat::Gray => {
                let mut new_buffer: Vec<u8> = Vec::new();
                new_buffer.reserve_exact(buffer.len());

                let count = buffer.len();
                let buf_ptr = buffer.as_ptr();
                let new_ptr = new_buffer.as_mut_ptr();
                // let src_ptr = buffer.as_ptr() as usize + buffer.len();
                // let dst_ptr = row.as_ptr() as usize;
                let new_range = new_ptr as usize ..(new_ptr as usize + count);
                let buf_range = buf_ptr as usize ..(buf_ptr as usize + count);

                assert!(
                    new_range.end <= buf_range.start || buf_range.end <= new_range.start,
                    "Source and destination memory ranges overlap, new vec {}-{} overlaps buffer {}-{} ", new_range.start, new_range.end, buf_range.start, buf_range.end
                );

                new_buffer.extend_from_slice(&buffer);

                GrayImage::from_raw(width as u32, height as u32, new_buffer)
                    .map(DynamicImage::ImageLuma8)
            }
```

then it says that they have the exact same ptr and address range: 
```
Panicked at /Users/sam/dev/pdfium-render/src/pdf/document/page/object/image.rs:529:17
Source and destination memory ranges overlap, new vec 1281272-3405272 overlaps buffer 1281272-3405272 
```

I'm not sure how to track down why that is, but it seems like promising lead.

**Comment by samsieber** [comment]
And I found it. We return a const pointer to the buffer, which then gets dropped. So the next time we allocate a vec, it happens to perfectly overlap that buffer:

```
    #[allow(non_snake_case)]
    fn FPDFBitmap_GetBuffer(&self, bitmap: FPDF_BITMAP) -> *const c_void {
        log::debug!("pdfium-render::PdfiumLibraryBindings::FPDFBitmap_GetBuffer()");

        let width = self.FPDFBitmap_GetWidth(bitmap);

        let height = self.FPDFBitmap_GetHeight(bitmap);

        let buffer_len = (width * height * PdfiumRenderWasmState::BYTES_PER_PIXEL) as usize;

        let state = PdfiumRenderWasmState::lock();

        let buffer_ptr = state
            .call(
                "FPDFBitmap_GetBuffer",
                JsFunctionArgumentType::Pointer,
                Some(vec![JsFunctionArgumentType::Pointer]),
                Some(&JsValue::from(Array::of1(&Self::js_value_from_bitmap(
                    bitmap,
                )))),
            )
            .as_f64()
            .unwrap() as usize;

        let buffer = state.copy_bytes_from_pdfium(buffer_ptr, buffer_len);

        buffer.as_ptr() as *const c_void
    }
```

Not sure what the fix is.

**Comment by samsieber** [comment]
When the two vecs were perfectly lined up in range, I was not actually running the reproduction, but had commented the line of code that made the image go corrupt. Running it on the reproduction gives me vecs that don't line up.

So, then it's definitely the FPDFBitmap_GetBuffer dropping the vec. If we can change the api of FPDFBitmap_GetBuffer to actually return a vec instead of a `*const c_void` that seems like it'd be the easiest solution. But I'm not sure what api constraints there would be in the codebase. I'll see what I can patch and get working locally.

**Comment by samsieber** [comment]
I submitted a quick-and-dirty patch. I don't expect it to be used, but mostly as a place to start.

**Comment by ajrcarey** [comment]
Thank you for all your detailed investigation into this, @samsieber - I really appreciate it. I concur with you that the problem starts here: https://github.com/ajrcarey/pdfium-render/blob/db87aac697a8dcdad1c7dd29e90f7fbc38f34ae3/src/bindings/wasm.rs#L6240

A pointer to `buffer` is returned, but the `buffer` itself is immediately dropped. The underlying reason for all this copying around of buffers is that `FPDFBitmap_GetBuffer()` will return a pointer into Pdfium's WASM linear memory address space; this pointer will not be valid within the Rust application's WASM linear memory address space, as WASM modules do not share memory spaces. (There is technically a way to do that, but it would require compiling the Pdfium WASM module with special flags, and since Pdfium and Rust use separate memory allocators with no knowledge of what the other is doing, it is virtually guaranteed that the two modules would be constantly clobbering one another's mallocs. Better to keep them separate.)

It is very hard to make this `FPDFBitmap_GetBuffer()` call memory safe. I see you have submitted a patch that would have it return a `Vec` instead of a pointer - that's an improvement, but isn't particularly idiomatic when compared to the rest of the API. I'm sort of wondering if it's better just to not provide that function at all when comiling to WASM. The consequences would be:

* Callers using the Pdfium API directly would have to use `FPDFBitmap_GetArray()` when compiling to WASM, an already-present WASM-specific function that works around the problem by returning a Javascript array slice of Pdfium's WASM linear memory address space. This has the additional advantage of avoiding the performance impact of needing to copy bitmaps from one WASM address space to another. 
* The `PdfPageImageObject::get_image_from_bitmap()` function (and similar) would need to be reworked to use `FPDFBitmap_GetArray()` when compiled to WASM.

There's no great solutions here, but part of me thinks it's better to have a clean break rather than try to paper over the cracks in the memory management. It also avoids the potential confusion of having `FPDFBitmap_GetBuffer()` return one data type when compiling to one target and a different data type when compiling to another - although I suppose it's subjective as to whether having two separate functions is any improvement over two separate return values.

**Comment by ajrcarey** [comment]
Having written all that, I'm now thinking in the following direction:

* Leave existing `FPDFBitmap_GetBuffer()` alone on targets that support it (currently, all targets except WASM)
* Remove `FPDFBitmap_GetBuffer()` for WASM target
* Add new `FPDFBitmap_GetBuffer_as_vec()` utility function for all targets that operates in the manner you have suggested in your PR
* Rename existing WASM-specific `FPDFBitmap_GetArray()` to `FPDFBitmap_GetBuffer_as_array()` for consistency

**Comment by samsieber** [comment]
That sound like a great path forward. Thanks for weighing in on this. I'd be happy to update my PR and try to bring it up to speed with your outline if you'd like, but it might not be until the week after Thanksgiving.

**Comment by ajrcarey** [comment]
That's fine, you've invested a lot of time into this and I can't ask you to do any more. I'll make the changes later on this week with a view to having them included in crate release 0.8.27.

**Comment by ajrcarey** [comment]
Implemented changes as discussed. Confirmed reproduction sample now behaves consistently. Updated README.

**Comment by ajrcarey** [comment]
Thank you again @samsieber for all your work on this issue. As there have been no further comments, and I believe the original problem is now resolved, I will shortly close this issue. Feel free to reopen if your testing shows the problem has not been correctly resolved.

**Comment by ajrcarey** [comment]
The fix will be released in crate version 0.8.27.
